pkg/experiment/calculator/experiment_calculator.go

```diff
@@ -0,0 +1,270 @@
+				return nil
+			}
+			gr := e.calcGoalResult(ctx, evalVc, goalVc, experiment.BaseVariationId)
+			e.appendVariationResult(ctx, timestamp, goalResult, gr.VariationResults)
+		}
+		experimentResult.GoalResults = append(experimentResult.GoalResults, goalResult)
+	}
+
+	return experimentResult
+}
+
+func (e ExperimentCalculator) listEnvironments(
+	ctx context.Context,
+) ([]*environment.Environment, error) {
+	listEnvironmentsRequest := environment.ListEnvironmentsRequest{
+		PageSize: 0,
+		Cursor:   "",
+	}
+	resp, err := e.environmentClient.ListEnvironments(ctx, &listEnvironmentsRequest)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Environments, err
+}
+
+func (e ExperimentCalculator) listExperiments(
+	ctx context.Context,
+	namespace string,
+) ([]*experiment.Experiment, error) {
+	stoppedAt := time.Now().Add(-2 * 24 * time.Hour).Unix()
+	whereParts := []mysql.WherePart{
+		mysql.NewFilter("deleted", "=", false),
+		mysql.NewFilter("environment_namespace", "=", namespace),
+		mysql.NewFilter("stopped_at", ">=", stoppedAt),
+		mysql.NewInFilter("status", []interface{}{
+			experiment.Experiment_WAITING,
+			experiment.Experiment_RUNNING,
+		}),
+	}
+	orders := []*mysql.Order{mysql.NewOrder("name", mysql.OrderDirectionAsc)}
+	experimentStorage := v2es.NewExperimentStorage(e.mysqlClient)
+
+	experiments, _, _, err := experimentStorage.ListExperiments(
+		ctx,
+		whereParts,
+		orders,
+		0,
+		0,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return experiments, nil
+}
+
+func (e ExperimentCalculator) getEvaluationCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentEvaluationCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	evaluationCount, err := e.eventCounterClient.GetExperimentEvaluationCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range evaluationCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) getGoalCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentGoalCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	goalCount, err := e.eventCounterClient.GetExperimentGoalCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range goalCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) calcGoalResult(
+	ctx context.Context,
+	evalVariationCounts,
+	goalVariationCounts map[string]*eventcounter.VariationCount,
+	baseVid string,
+) *eventcounter.GoalResult {
+	var (
+		goalResult = &eventcounter.GoalResult{}
+		vids       = make([]string, 0, len(goalVariationCounts))
+	)
+	for vid, _ := range goalVariationCounts {
+		vids = append(vids, vid)
```

Comment:
The vids var is not being used. Is this function unfinished?

Reply:
Yes, I combed through the logic and listed the functions that need to be implemented of the Python Calculator. So the function not be fully implemented yet.